mixnet/robustness.py

```diff
@@ -27,37 +24,77 @@ class Robustness:
 
     def __init__(
         self,
-        mixnode_candidates: List[MixNode],
-        mixnet_topology_size: MixnetTopologySize,
+        config: RobustnessConfig,
         mixnet: Mixnet,
     ) -> None:
-        assert mixnet_topology_size.num_total_mixnodes() <= len(mixnode_candidates)
-        self.mixnode_candidates = mixnode_candidates
-        self.mixnet_topology_size = mixnet_topology_size
-        self.mixnet = mixnet
+        self.__config = config
+        self.__mixnet = mixnet
 
     def set_entropy(self, entropy: bytes) -> None:
         """
         Given a entropy received, build a new topology and send it to mixnet.
+        In v1, this doesn't change any mixnet config except topology.
 
         In real implementations, this method should be a long-running task, consuming entropy periodically.
         Here in the spec, this method has been simplified as a setter, assuming the single-thread test environment.
         """
-        topology = self.build_topology(entropy)
-        self.mixnet.set_topology(topology)
+        self.__config.mixnet.mixnet_layer_config.topology = self.build_topology(
+            self.__config.mixnet.mixnode_candidates,
+            self.__config.mixnet.topology_size,
+            entropy,
+        )
+        self.__mixnet.set_config(self.__config.mixnet.mixnet_layer_config)
 
-    def build_topology(self, entropy: bytes) -> MixnetTopology:
+    @staticmethod
+    def build_topology(
+        mixnode_candidates: List[MixNodeInfo],
+        mixnet_topology_size: MixnetTopologySize,
+        entropy: bytes,
+    ) -> MixnetTopology:
         """
         Build a new topology deterministically using an entropy and a given set of candidates.
         """
-        shuffled = FisherYates.shuffle(self.mixnode_candidates, entropy)
-        sampled = shuffled[: self.mixnet_topology_size.num_total_mixnodes()]
+        shuffled = FisherYates.shuffle(mixnode_candidates, entropy)
+        sampled = shuffled[: mixnet_topology_size.num_total_mixnodes()]
 
         layers = []
-        for layer_id in range(self.mixnet_topology_size.num_layers):
-            start = layer_id * self.mixnet_topology_size.num_mixnodes_per_layer
-            layer = sampled[
-                start : start + self.mixnet_topology_size.num_mixnodes_per_layer
-            ]
+        for layer_id in range(mixnet_topology_size.num_layers):
+            start = layer_id * mixnet_topology_size.num_mixnodes_per_layer
+            layer = sampled[start : start + mixnet_topology_size.num_mixnodes_per_layer]
             layers.append(layer)
         return MixnetTopology(layers)
+
+
+@dataclass
+class RobustnessConfig:
+    """In v1, the robustness layer manages configs only for the mixnet layer."""
```

Comment:
We have discussed this. And I am not really happy with the naming as well. But, I think this is a coordination layer **just for the mixnet itself**. Probably for the best not to share responsibilities with any other thing.

Reply:
I've had similar thoughts, as we discussed in https://github.com/logos-co/nomos-node/pull/570#discussion_r1469895805. In my honest opinion, if this is only for mixnet, we don't really need to define this as a separate component (layer, class, service, or whatever). If it's better to have this independently for later uses, we need to clarify what this component will be responsible for other components as well as mixnet. I think we need to discuss about this more with @madxor here.

Reply:
We're investing much time for this topic but still not making a conclusion yet. I just made a separate PR to boost our conclusion. Please have a look https://github.com/logos-co/nomos-specs/pull/61

Reply:
I'm merging this PR before it's getting hard to track all changes in this PR. This discussion will be continued in #61.